# Random Forest
This Jupyter Notebook will examine the words of the same dataset as used on the naive bayes models, to try and predict the gender based on the wording used in resumes and twitter texts.

In [1]:
import pandas as pd
#Importing the same dataset (resumes + twittertext)
df = pd.read_csv('./resumes300.csv')
df

gender                                               text
0      female  PROFILE Fund accountant with nearly  years of ...
1        male  Fund Accoutant Sep  Present Citco Fund Service...
2      female  Professional Experience Citco Fund Services Si...
3        male  PROFESSIONAL EXPERIENCE Conifer Financial Serv...
4        male  QUALIFICATION SUMMARY Results driven achiever ...
...       ...                                                ...
13223  female  @lookupondeath ...Fine, and I'll drink tea too...
13224    male  Greg Hardy you a good player and all but don't...
13225    male  You can miss people and still never want to se...
13226  female  @bitemyapp i had noticed your tendency to pee ...
13227  female  I think for my APUSH creative project I'm goin...

[13228 rows x 2 columns]

## the model

In [2]:
from sklearn.feature_extraction.text import CountVectorizer #The CountVectorizer object

text = df['text'].values.astype('U') #Taking the text from the df. We need to convert it to Unicode
vect = CountVectorizer(stop_words='english') #Create the CV object, with English stop words
vect = vect.fit(text) #We fit the model with the words from the review text
vect
feature_names = vect.get_feature_names() #Get the words from the vocabulary
print(f"There are {len(feature_names)} words in the vocabulary. A selection: {feature_names[500:520]}")
docu_feat = vect.transform(text) #The transform method from the CountVectorizer object creates the matrix
docu_feat

There are 36090 words in the vocabulary. A selection: ['accident', 'accidental', 'accidentally', 'accidents', 'acciona', 'accommodation', 'accompaniment', 'accompany', 'accompanying', 'accomplish', 'accomplished', 'accomplishing', 'accomplishment', 'accomplishments', 'accordance', 'accorded', 'according', 'accordingly', 'account', 'accountabilities']


<13228x36090 sparse matrix of type '<class 'numpy.int64'>'
	with 204930 stored elements in Compressed Sparse Row format>

In [6]:
from sklearn.feature_extraction.text import TfidfTransformer
tfidf_transformer = TfidfTransformer()
X_train_tf = tfidf_transformer.fit_transform(docu_feat)
X_train_tf.eliminate_zeros()
X_train_tf.shape

(13228, 36090)

In [7]:
y = df['gender'] #We need to take out the gender as our Y-variable
X = X_train_tf  #this slices the dataframe to include all rows I need
X 

<13228x36090 sparse matrix of type '<class 'numpy.float64'>'
	with 204930 stored elements in Compressed Sparse Row format>

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

## how well is the model performing?

In [ ]:
from sklearn.ensemble import RandomForestClassifier
train_data_features=X_train.toarray()
forest = RandomForestClassifier(n_estimators = 100) 
forest = forest.fit( train_data_features, y_train)
forest

In [13]:
from sklearn.metrics import accuracy_score
result = forest.predict(X_test)
print(accuracy_score(y_test, result))

0.5933484504913077


An accuracy of 0.59 is still too low to make accurate predictions on a variable that only has 2 outcomes. This test is quite possibly failing due to lack of relevant texts (resumes) and the twitter data makes for noise in the dataset. Next tests should include more relevant data of resume texts.